servus/serializable.cpp

```diff
@@ -23,37 +23,128 @@
 
 namespace servus
 {
+
+class Serializable::Impl
+{
+public:
+    void registerChangedCallback( const Serializable::ChangeFunc& func )
+    {
+        changed = func;
+    }
+
+    void registerDeserializedCallback(
+                                     const Serializable::DeserializeFunc& func )
+    {
+        deserialized = func;
+    }
+
+    void registerSerializeCallback( const Serializable::SerializeFunc& func )
+    {
+        serialize = func;
+    }
+
+    void notifyChanged() const
+    {
+        if( changed )
+            changed();
+    }
+
+    void notifyDeserialized() const
+    {
+        if( deserialized )
+            deserialized();
+    }
+
+    void notifySerialize() const
+    {
+        if( serialize )
+            serialize();
+    }
+
+    Serializable::ChangeFunc changed;
+    Serializable::DeserializeFunc deserialized;
+    Serializable::SerializeFunc serialize;
+};
+
+Serializable::Serializable()
+    : _impl( new Serializable::Impl( ))
+{}
+
+Serializable::~Serializable()
+{
+    delete _impl;
+}
+
 uint128_t Serializable::getTypeIdentifier() const
 {
     return make_uint128( getTypeName( ));
 }
 
-Serializable::ChangeFunc Serializable::setUpdatedFunction(
-    const ChangeFunc& func )
+bool Serializable::fromBinary( const Data& data )
+{
+    return fromBinary( data.ptr.get(), data.size );
+}
+
+bool Serializable::fromBinary( const void* data, const size_t size )
+{
+    if( _fromBinary( data, size ))
+    {
+        _notifyDeserialized();
+        return true;
+    }
+    return false;
+}
+
+Serializable::Data Serializable::toBinary() const
+{
+    _notifySerialize();
+    return _toBinary();
+}
+
+bool Serializable::fromJSON( const std::string& json )
+{
+    if( _fromJSON( json ))
+    {
+        _notifyDeserialized();
```

Comment:
Absolutely, this is the right place to call _notifyDeserialized().
However, as you pointed out, the setters in ZeroBuf also call notifyChanged(), which may not the desired effect. Let me think about this.